### ISSUE TITLE
Having an ML agent play a game - Training/Config issues

### ISSUE THREAD

**Issue by Smileynator** [Unity-Technologies/ml-agents#2552]
So i am just going to try and explain my issues as best i can and am allowed to with as much details as possible.

Currently i have a Archero clone type game, and i am trying to make a ML-agent play the role of the player, to easily try out different game balancing without having to keep testing the game all manually.

### Gameplay / Environment
The game is a level based top down shooter type. You enter the level, the player autofires in line of sight, some enemies are melee some are shooting back. The goal is to kill all enemies in the level, and then proceed to the next level gate. The losing condition is death.

The agent gets 20 seconds to complete a level, any damage dealt to any enemy, resets this timer to 20 seconds to ensure boss/difficult levels still get enough time to complete properly, but we don't get massive time stuck in episodes that are going nowhere.
- If the timer runs out the Episode ends.
- If the agent dies, the episode also ends.
- Timescale is not relevant so during training everything just runs at 8x speed over 4 build clients.
- Levels in the game are in a fixed order and fixed layout, but every level is unique in exit position, enemies, and inner layout of obstacles.

### Input neurons (32)
- (2) player delta velocity X & Y (-1/+1) Not normalized but never bigger than (-0.1/+0.1)
- (1) player HP (0/+1) Normalized with MAX HP
- (1) player Y Rotation (0/+1) Normalized
- (1) player Forward Distance CapsuleCast (0/+1) Normalized with max cast distance
- (3) player Directional (back, left, right) Distance Raycast (0/+1) Normalized with max distance 
- (2) exit location X & Y (-1/+1) Normalized relative to player position and capped by max level size
- (1) enemies alive (0/+1) boolean statement if ANY enemy is alive
- (1) enemies alive (0/+1) normalized amount of enemies still alive (current/25)
- (10) 5 closest Enemies X & Y (-1/+1) Normalized relative to player position and capped by max level size
- (10) 5 closest enemy projectiles X & Y (-1/+1) Normalized relative to player position and capped by max level size

All neurons if no info is available default to 0f. Except for closest enemy and projectiles, they default to -1f, because i thought it might be clearer to the neural net than differentiating 0.01f with 0f as enemy present and in your face, or no enemy present. Not sure if that helps though.

### Output neurons (2)
-  player joystick X & Y axis (-1/+1)

### Unity settings Agent
Max Step = 0
Reset on Done = true
On Demand Decision = false
Decision Interval = 3

### Unity settings Brain
Vector Obs Size = 32
Stacked Vectors = 4 (projectiles fly, enemies move, made sense?)
Vector Action Type = Continuous
Vector Action Size = 2

### Unity rewards and punishment moments
- Closing in on the exit gate (0.005 per frame if closer than any previous frame) (closest value reset per level, so no back/forward abuse)
- If all enemies are dead, and not making progress to the gate (-0.025 once every 2 seconds)
- If enemies are alive, and not hitting any enemy (-0.075 once every 3 seconds)
- If level exit reached (0.25)
- If any amount of damage taken (-0.25)
- If enemy damaged for any amount (0.01)
- If enemy killed (0.1)
- If the timer expires or player dies (episode end) (-0.4)

The exit gate one, used to be only for the first 5 levels, but that seemed to cause issues in longer training sessions. The punishment for not reaching exit is there to attempt learning how to reach the exit the right way (punish it getting stuck). Punishing of enemies not hit is to attempt it to actually go out in the level and hunt the last enemies (seems to help better than promoting a kill or damaging, because then it rather just hug the door and hope enemies come to him)

### Training Config (recent changes in brackets)
>     trainer: ppo
>     batch_size: 1024 (was 2048)
>     beta: 2.5e-3 (was 1.0e-3 i think)
>     buffer_size: 10240 (was 20280)
>     epsilon: 0.21
>     hidden_units: 128 (assuming 32 input neurons * 4 stacked vectors?)
>     lambd: 0.95
>     learning_rate: 0.5e-4 (changed, no idea what old value was)
>     max_steps: 5.0e7 (long because of overnight training, and every 3 frames = 1 step)
>     memory_size: 256
>     normalize: false (recently set to false, i normalize all input myself anyway)
>     num_epoch: 3
>     num_layers: 2 (was 3)
>     time_horizon: 256
>     sequence_length: 32
>     summary_freq: 25000 (was 10000)
>     use_recurrent: false
>     vis_encode_type: simple
>     reward_signals: 
>         extrinsic:
>             strength: 1.0
>             gamma: 0.99

### The problem we have
The agents have a lot of issues with orientation and movement. They seem to get the point (dodge bullets, don't run into enemies, run to the exit etc.) but once it becomes complicated (agent gets overwhelmed by enemies, player is evading but hits a wall in the process, player gets to the side of the exit, but somehow doesn't figure out how to walk around the open door into the exit etc.)

They don't seem to be able to figure out these complicated scenarios. So i added more input (back/side raycasts and player rotation) in order to help it out, but it does not seem to be making it any better even over long long training sessions they still keep falling for those pitfalls even though in other aspects (bullet dodging is quite late in the learning process) seem to have no problem learning new things that are not this navigation issue.

I think i am missing something in regards of what makes the Agents "blind" or somehow not able to solve these issues, and i might not have the optimal learning settings either, but any tweaks to the later seem to not improve the settings.

I am pretty sure i have debugged all the input/output neurons to be correct and as intended, and all reward functions happen as intended. Fixing some of these actually caused some mayor improvements before, so i make sure to check then regularly after changes.

Attached is out TensorFlow dashboard highlighting some of my recent attempts to fix this problem. In our case the "session length" is the real value we want to get higher, because the further it gets through the levels the better. The 2 longer learning attempts were still without the improved input and with 30 second per level (no timer resets on enemies attacked) and old Training configs.

![TensorFlow](https://user-images.githubusercontent.com/3433068/64856368-1b7e3980-d622-11e9-8faf-78bb99a43add.PNG)

I hope someone sees some things that can point out what i am doing wrong. I think i know a lot of basic information and guides. A lot of general info. But not a lot of reference material or knowhow to properly configure my Config file for optimal learning. And i seem to miss some logic as to why some things are not learned right now when in my eyes all the parameters are there.

If you have any more questions or details, besides screenshots of the game, i will try to help provide the information as well as i can.

Thanks in advance!

**Comment by Smileynator** [comment]
I have so far not made any significant improvements. I let my ML play out over the weekend to see where it would end up.

It seems to have significantly worse than before the tweaks i have been doing that should be all documentation and advice i have read online, have improved the behavior of the neural net. (less over-fitting etc.)

Of course this could imply that before it was over-fitting to such a degree that it somehow remembered it's surroundings and just fit level specific knowledge in that memory instead of actually learning the tasks proper?

Sadly it does not help me improve the neural net in any way currently.

Here is the longer run data:
![weekend](https://user-images.githubusercontent.com/3433068/65043135-9d3fd100-d95a-11e9-97b7-fc0bbcd27359.PNG)

PS: I saw a thread here about games with ML agents, would this apply to that? I am doing this as a worker for a company after all.

**Comment by Smileynator** [comment]
After some more study on what the behavior is, it seems that when the agent starts to progress through rooms with more enemies and obstacles in them. That the agent learns about killing enemies and the reward that gets them. But promptly stars forgetting/not caring about actually going through to the end of the room (even though that is the biggest reward they will get in their sessions.

Is there any good tips to make multiple objectives work? Is it just as simple as "only reward the end goal, and let it figure it out even if it takes 100 million attempts" or is there a way to get them to not forget about their original goals?
Or is the fact that i am spoonfeeding it little intermediate rewards/punishments actually just confusing to the agent?
Kind of at a loss on what is best to do here in order to not confuse the agent on what i want it to do. Any advice?

**Comment by awjuliani** [comment]
When it comes to attempting to balance multiple objectives, the simpler the reward function, the better. So if the agent has to kill enemies to get to a goal, then just rewarding the goal *should* eventually be enough to have the agent learn that it needs to also kill the enemies in order to achieve that goal. Of course the more steps you put in between the start of the episode and the sparse reward at the end the harder it will become for the agent to learn from just that signal.

**Comment by Smileynator** [comment]
@awjuliani 
Making it that way, i figure it would take quite a while for it to randomly get to the goal and learn, seeing how it could take like 20 seconds for 1 reward to happen (ingame time)

What is a good way to "hand hold" it through the steps?
Could it also be that the fact that the level progression is static (always start at level 1, and proceed until dead) actually makes it next to impossible for it to properly train all the way?
Considering that level 6 or 7 and later level 12 still introduce new mechanics to the actor?

**Comment by awjuliani** [comment]
One thing to try is concurrently training on instances of all the levels, to prevent the issue you've brought up.

**Comment by Smileynator** [comment]
If concurrent isn't possible (i can possibly run 4 instances max) what would be the best fix after that? Would randomizing the levels in a way that all levels are seen an equal amount of times over the total learning time work?

**Comment by awjuliani** [comment]
Yes, I would recommend randomizing the level order.

**Comment by Smileynator** [comment]
@awjuliani This seems to clear up most issues. It takes a lot more time to learn this way though. It still seems to have some issues with navigation around obstacles. In general, how much steps would it need to train for to learn something as complex as a game like i described above? Because i have seen machine learning run into millions of steps before anything even started learning properly.

Also to restate the thing from before, would my usecase be relevant to the request of "machine learning used in games"? that i saw in the issue list?

**Comment by firepro20** [comment]
@awjuliani does this also holds true for a static level where an agent has to collect multiple health items, but only gets a goal at the end after collecting all of them? I have just removed the reward for collecting one each time, and just provided a goal/rewards when all are collected. So far it looks good. However since I am using continuous space I cannot make use of memory in order to remember which ones have been collected and the ones that aren't. Any suggestions?
